src/components/icon/icon-registry.ts

```diff
@@ -0,0 +1,370 @@
+   * returns it. Returns null if no matching element is found.
+   */
+  private _extractSvgIconFromSet(
+      iconSet: SVGElement, iconName: string, config: SvgIconConfig): SVGElement {
+    const iconNode = iconSet.querySelector('#' + iconName);
+    if (!iconNode) {
+      return null;
+    }
+    // If the icon node is itself an <svg> node, clone and return it directly. If not, set it as
+    // the content of a new <svg> node.
+    if (iconNode.tagName.toLowerCase() == 'svg') {
+      return this._setSvgAttributes(<SVGElement>iconNode.cloneNode(true), config);
+    }
+    // createElement('SVG') doesn't work as expected; the DOM ends up with
+    // the correct nodes, but the SVG content doesn't render. Instead we
+    // have to create an empty SVG node using innerHTML and append its content.
+    // http://stackoverflow.com/questions/23003278/svg-innerhtml-in-firefox-can-not-display
+    const svg = this._svgElementFromString('<svg></svg>');
+    // Clone the node so we don't remove it from the parent icon set element.
+    svg.appendChild(iconNode.cloneNode(true));
+    return this._setSvgAttributes(svg, config);
+  }
+
+  /**
+   * Creates a DOM element from the given SVG string.
+   */
+  private _svgElementFromString(str: string): SVGElement {
+    // TODO: Is there a better way than innerHTML? Renderer doesn't appear to have a method for
+    // creating an element from an HTML string.
+    const div = document.createElement('DIV');
+    div.innerHTML = str;
+    const svg = <SVGElement>div.querySelector('svg');
+    if (!svg) {
+      throw new MdIconSvgTagNotFoundException();
+    }
+    return svg;
+  }
+
+  /**
+   * Sets the default attributes for an SVG element to be used as an icon.
+   */
+  private _setSvgAttributes(svg: SVGElement, config: SvgIconConfig): SVGElement {
+    if (!svg.getAttribute('xmlns')) {
+      svg.setAttribute('xmlns', 'http://www.w3.org/2000/svg');
+    }
+    svg.setAttribute('fit', '');
+    svg.setAttribute('height', '100%');
+    svg.setAttribute('width', '100%');
+    svg.setAttribute('preserveAspectRatio', 'xMidYMid meet');
+    svg.setAttribute('focusable', 'false'); // Disable IE11 default behavior to make SVGs focusable.
+    return svg;
+  }
+
+  /**
+   * Returns an Observable which produces the string contents of the given URL. Results may be
+   * cached, so future calls with the same URL may not cause another HTTP request.
+   */
+  private _fetchUrl(url: string): Observable<string> {
+    // Store in-progress fetches to avoid sending a duplicate request for a URL when there is
+    // already a request in progress for that URL. It's necessary to call share() on the
+    // Observable returned by http.get() so that multiple subscribers don't cause multiple XHRs.
+    if (this._inProgressUrlFetches.has(url)) {
+      return this._inProgressUrlFetches.get(url);
+    }
+    const req = this._http.get(url)
+        .map((response) => response.text())
```

Comment:
You probably want a `flatMap` here.

Also, http can cache IIRC.  Let me check the API.